packages/flight_controller/src/flight_controller_node.py

```diff
@@ -0,0 +1,381 @@
+        # 90 degrees, the body-frame z-axis is perpendicular to the force of
+        # gravity, so, as if the drone were in free-fall (which was roughly
+        # confirmed experimentally), the IMU reads -9.8 m/s^2 along the z-axis.
+        g = 9.8
+        lin_acc_x_drone_body = lin_acc_x_drone_body + g*np.sin(roll)*np.cos(pitch)
+        lin_acc_y_drone_body = lin_acc_y_drone_body + g*np.cos(roll)*(-np.sin(pitch))
+        lin_acc_z_drone_body = lin_acc_z_drone_body + g*(1 - np.cos(roll)*np.cos(pitch))
+
+        # calculate the angular velocities of roll, pitch, and yaw in rad/s
+        time = rospy.Time.now()
+        dt = time.to_sec() - self.time.to_sec()
+        dr = roll - previous_roll
+        dp = pitch - previous_pitch
+        dh = heading - previous_heading
+        angvx = self.near_zero(dr / dt)
+        angvy = self.near_zero(dp / dt)
+        angvz = self.near_zero(dh / dt)
+        self.time = time
+
+        # Update the imu_message:
+        # header stamp
+        self.imu_message.header.stamp = time
+        # orientation
+        self.imu_message.orientation.x = quaternion[0]
+        self.imu_message.orientation.y = quaternion[1]
+        self.imu_message.orientation.z = quaternion[2]
+        self.imu_message.orientation.w = quaternion[3]
+        # angular velocities
+        self.imu_message.angular_velocity.x = angvx
+        self.imu_message.angular_velocity.y = angvy
+        self.imu_message.angular_velocity.z = angvz
+        # linear accelerations
+        self.imu_message.linear_acceleration.x = lin_acc_x_drone_body
+        self.imu_message.linear_acceleration.y = lin_acc_y_drone_body
+        self.imu_message.linear_acceleration.z = lin_acc_z_drone_body
+
+    def update_battery_message(self):
+        """
+        Compute the ROS battery message by reading data from the board.
+        """
+        # extract voltage, current
+        self.board.getData(MultiWii.ANALOG)
+        # Update Battery message:
+        self.battery_message.voltage = self.board.analog['vbat'] * 0.10
+        self.battery_message.current = self.board.analog['amperage']
+
+    def update_command(self):
+        ''' Set command values if the mode is ARMED or DISARMED '''
+        if self.curr_mode == 0: #disarmed
+            self.command = cmds.disarm_cmd
+        elif self.curr_mode == 1: #armed
+            if self.prev_mode == 0: #disarmed
+                self.command = cmds.arm_cmd
+            elif self.prev_mode == 1: #armed
+                self.command = cmds.idle_cmd
+
+    # Helper Methods:
+    #################
+    def getBoard(self):
+        """ Connect to the flight controller board """
+        # (if the flight cotroll usb is unplugged and plugged back in,
+        #  it becomes .../USB1)
+        try:
+            board = MultiWii('/dev/ttyUSB0')
+        except SerialException:
+            try:
+                board = MultiWii('/dev/ttyUSB1')
+            except SerialException:
```

Comment:
I don't understand this comment, or rather how else it would work.

Reply:
@amackeith  in linux, you can define device rules to map a known device to a given device path. Devices are identified by the pair (VID, PID), where VID = Vendor ID (each company has one ID assigned) and PID = Product ID (each company assigns one ID to each product they make). Udev allows you to create rules like:

> ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0461", ATTRS{idProduct}=="4d81", SYMLINK+="myusb"

which means:

> if a new device with `VID=0461` and `PID=4d81` is attached to the `BUS=usb`, create a symlink to it at `/dev/myusb`

To learn more, look at this example:
https://askubuntu.com/questions/783357/create-a-symlink-in-dev-using-etc-udev-rules

Reply:
@amackeith 
can you try and get this to work on the drone?
dev rules cannot be defined inside Docker, so, you will need to put this in `/lib/udev/rules.d/` of your raspberry pi to work. Once you get it to work, please, share the rule file with me so that I can put in the `init_sd_card` procedure so that we flash it together with the OS.